components/sdmmc/sdmmc_cmd.c

```diff
@@ -263,61 +279,214 @@ esp_err_t sdmmc_card_init(const sdmmc_host_t* config, sdmmc_card_t* card)
     }
 
     if (card->is_mem) {
-        /* SDSC cards support configurable data block lengths.
-         * We don't use this feature and set the block length to 512 bytes,
-         * same as the block length for SDHC cards.
-         */
-        if ((card->ocr & SD_OCR_SDHC_CAP) == 0) {
-            err = sdmmc_send_cmd_set_blocklen(card, &card->csd);
-            if (err != ESP_OK) {
-                ESP_LOGE(TAG, "%s: set_blocklen returned 0x%x", __func__, err);
-                return err;
-            }
-        }
-        /* Get the contents of SCR register: bus width and the version of SD spec
-         * supported by the card.
-         * In SD mode, this is the first command which uses D0 line. Errors at
-         * this step usually indicate connection issue or lack of pull-up resistor.
-         */
-        err = sdmmc_send_cmd_send_scr(card, &card->scr);
-        if (err != ESP_OK) {
-            ESP_LOGE(TAG, "%s: send_scr (1) returned 0x%x", __func__, err);
-            return err;
-        }
-    }
-
-    if (card->is_mem) {
-        /* If the host has been initialized with 4-bit bus support, and the card
-         * supports 4-bit bus, switch to 4-bit bus now.
-         */
-        if ((card->host.flags & SDMMC_HOST_FLAG_4BIT) &&
-            (card->scr.bus_width & SCR_SD_BUS_WIDTHS_4BIT)) {
-            ESP_LOGD(TAG, "switching to 4-bit bus mode");
-            err = sdmmc_send_cmd_set_bus_width(card, 4);
-            if (err != ESP_OK) {
-                ESP_LOGE(TAG, "set_bus_width failed");
-                return err;
-            }
-            err = (*config->set_bus_width)(config->slot, 4);
-            if (err != ESP_OK) {
-                ESP_LOGE(TAG, "slot->set_bus_width failed");
-                return err;
-            }
-        }
-
-        /* Wait for the card to be ready for data transfers */
-        uint32_t status = 0;
-        while (!is_spi && !(status & MMC_R1_READY_FOR_DATA)) {
-            // TODO: add some timeout here
-            uint32_t count = 0;
-            err = sdmmc_send_cmd_send_status(card, &status);
-            if (err != ESP_OK) {
-                return err;
-            }
-            if (++count % 16 == 0) {
-                ESP_LOGV(TAG, "waiting for card to become ready (%d)", count);
-            }
-        }
+		if (card->host.flags & SDMMC_HOST_MMC_CARD) {	//MMC CARD
+			/* sdmmc_mem_mmc_init */
+			int width, value;
+			int card_type;
+			int speed = 20000;
+			uint8_t powerclass = 0;
+			//malloc(size) on esp32 is same as heap_caps_malloc(size, MALLOC_CAP_8BIT)
+			//!!!remember to free(ext_csd) before all return-s in this block !!!
+			#define EXT_CSD_SIZE 512
+			uint8_t* ext_csd = malloc(EXT_CSD_SIZE);		
```

Comment:
If you are passing this buffer to the host driver, it might need to be in DMA-capable memory, so `heap_caps_malloc` with `MALLOC_CAP_DMA` would be better.

Reply:
done